apps/client/src/liquidityVenues/midas/index.ts

```diff
@@ -0,0 +1,329 @@
+  }
+
+  private redemptionVault(token: Address, chainId: number) {
+    return midasConfigs[chainId]![token]!.instantRedemptionVault;
+  }
+
+  previewRedeemInstant(params: PreviewRedeemInstantParams) {
+    const feeData = this._calcAndValidateRedeem(params);
+    if (!feeData) return undefined;
+
+    if (!this._requireAndUpdateLimit(params, feeData.amountMTokenWithoutFee)) return undefined;
+
+    const usdData = this._convertMTokenToUsd(params, feeData.amountMTokenWithoutFee);
+
+    if (!usdData) return undefined;
+
+    const tokenData = this._convertUsdToToken(params, usdData.amountUsd);
+
+    if (!tokenData) return undefined;
+
+    return {
+      amountTokenOutWithoutFee: this._truncate(
+        (feeData.amountMTokenWithoutFee * usdData.mTokenRate) / tokenData.tokenRate,
+        params.tokenOutDecimals,
+      ),
+      feeAmount: feeData.feeAmount,
+    };
+  }
+
+  private _calcAndValidateRedeem(params: PreviewRedeemInstantParams) {
+    if (params.minAmount > params.amountMTokenIn) return undefined;
+
+    const feeAmount = this._getFeeAmount(params);
+
+    return params.amountMTokenIn > feeAmount
+      ? { feeAmount, amountMTokenWithoutFee: params.amountMTokenIn - feeAmount }
+      : undefined;
+  }
+
+  private _getFeeAmount(params: PreviewRedeemInstantParams) {
+    if (params.waivedFeeRestriction) return 0n;
+
+    const feePercent = MathLib.min(
+      params.tokenOutConfig.fee + params.instantFee,
+      ONE_HUNDRED_PERCENT,
+    );
+
+    return (params.amountMTokenIn * feePercent) / ONE_HUNDRED_PERCENT;
+  }
+
+  private _requireAndUpdateLimit(params: PreviewRedeemInstantParams, amount: bigint) {
+    return params.dailyLimits + amount <= params.instantDailyLimit;
+  }
+
+  private _convertMTokenToUsd(params: PreviewRedeemInstantParams, amount: bigint) {
+    if (amount === 0n || params.mTokenRate === 0n) return undefined;
+
+    return {
+      amountUsd: (amount * params.mTokenRate) / 10n ** 18n,
+      mTokenRate: params.mTokenRate,
+    };
+  }
+
+  private _convertUsdToToken(params: PreviewRedeemInstantParams, amountUsd: bigint) {
+    if (amountUsd === 0n) return undefined;
+
+    const tokenRate = params.tokenOutConfig.stable ? params.STABLECOIN_RATE : params.tokenOutRate;
+
+    if (tokenRate === 0n) return undefined;
+
+    return {
+      amountToken: (amountUsd * 10n ** 18n) / tokenRate,
+      tokenRate,
+    };
+  }
+
+  private _truncate(value: bigint, decimals: bigint) {
+    return this._convertToBase18(this._convertFromBase18(value, decimals), decimals);
+  }
+
+  private _convertFromBase18(originalAmount: bigint, decidedDecimals: bigint) {
+    return this._convert(originalAmount, 18n, decidedDecimals);
+  }
+
+  private _convertToBase18(originalAmount: bigint, originalDecimals: bigint) {
+    return this._convert(originalAmount, originalDecimals, 18n);
+  }
+
+  private _convert(originalAmount: bigint, originalDecimals: bigint, decidedDecimals: bigint) {
```

Comment:
nit: naming -- what do you think of something like `_rescale(amount: bigint, oldDecimals: bigint, newDecimals: bigint)`